src/ogdf/basic/graph_generators/operations.cpp

```diff
@@ -246,4 +251,83 @@ void rootedProduct(const Graph& G1, const Graph& G2, Graph& product, NodeMap& no
 	});
 }
 
+void complement(Graph& G, bool directional, bool allow_self_loops) {
+	for (node n1 : G.nodes) {
+		for (node n2 : G.nodes) {
+			if (n1->index() > n2->index()) {
+				continue;
+			}
+			if (!allow_self_loops && n1->index() == n2->index()) {
+				continue;
+			}
+
+			edge e = G.searchEdge(n1, n2, true);
```

Comment:
`searchEdge` has linear runtime in the degree, so this code has super-quadratic runtime. Either use [`AdjacencyOracle`](https://ogdf.github.io/doc/ogdf/classogdf_1_1_adjacency_oracle.html#ad6f1706cd11b76bb126c662411c1f063) or even better a `NodeSet` initialized with all neighbors of `n1` before the `n2` loop.